File: modules/custom-audio-module/src/CustomAudioModule.ts
import { requireNativeModule } from 'expo';

// This call loads the native module object from the JSI.
const CustomAudioModule = requireNativeModule('CustomAudioModule');

// AudioSource type
export type AudioSource = string;

// AudioPlayer interface
export interface AudioPlayer {
  /**
   * Boolean value indicating whether the player is currently playing.
   */
  playing: boolean;

  /**
   * Boolean value indicating whether the player is finished loading.
   */
  isLoaded: boolean;

  /**
   * Start playing audio.
   */
  play(): void;

  /**
   * Stops/pauses the player.
   */
  stop(): void;

  /**
   * Load an audio source.
   */
  loadAudio(source: AudioSource): Promise<void>;

  /**
   * Remove the player from memory to free up resources.
   */
  remove(): void;
}

// Export the AudioPlayer constructor from the native module
export namespace AudioPlayer {
  export const create = (): AudioPlayer => {
    return new (CustomAudioModule.AudioPlayerObject as new () => AudioPlayer)();
  };
}

File: modules/custom-audio-module/ios/CustomAudioModule.swift
import ExpoModulesCore

public class CustomAudioModule: Module {
  public func definition() -> ModuleDefinition {
    Name("CustomAudioModule")
    
    // Export the AudioPlayer SharedObject class
    Class(AudioPlayerObject.self) {
      Constructor { () -> AudioPlayerObject in
        return AudioPlayerObject()
      }
      
      // Properties - bound natively for performance
      Property("playing") { (audioPlayer: AudioPlayerObject) -> Bool in
        return audioPlayer.audioPlayer?.isPlaying ?? false
      }
      
      Property("isLoaded") { (audioPlayer: AudioPlayerObject) -> Bool in
        return audioPlayer.audioPlayer != nil
      }
      
      // Functions
      AsyncFunction("loadAudio") { (audioPlayer: AudioPlayerObject, source: String) in
        try await audioPlayer.loadAudio(source: source)
      }
      
      Function("play") { (audioPlayer: AudioPlayerObject) in
        try audioPlayer.play()
      }
      
      Function("stop") { (audioPlayer: AudioPlayerObject) in
        audioPlayer.stop()
      }
      
      Function("remove") { (audioPlayer: AudioPlayerObject) in
        audioPlayer.remove()
      }
    }
  }
}
